records/2024-05-23.scalability.short-code-implementation.md

```diff
@@ -0,0 +1,107 @@
+# Short Code Implementation
+
+Date: 2025-05-23
+
+## Status
+
+**DRAFT**.
+
+## Context
+
+Notify was awarded a General Purpose Short Code by the Canadian Telecom Regulators. This ADR goes through when an SMS will be sent through a Short Code or Long Code and the decision behind why a certain method will be chosen.
+
+## Implementation Goals
+1. Assign SMS templates to send through our generic shortcode.
+1. Assign any template to high, normal, or bulk priority automatically (SMS & Email)
+1. Design a way to categorize templates
+1. Override to force a certain template to go through a certain flow
+1. The ability for users to select “Other” and enter their own use case for Template Categories
+
+## Goals to be addressed by another ADR
+1. Figure out dynamic and static pools of numbers
+1. Decide on using AWS Pinpoint or AWS SNS
+Usage of a pool of codes/ dedicated long codes
+
+## Current Context and Assumptions
+
+1. **We assume that ALL Priority SMS Messages will be sent using the Short Code**
+1. For the purposes of this ADR, assume SMSs are sent to a send_sms client call. Under the hood the send_sms call might utilize SMS or Pinpoint.
+1. Currently, we have 3 priority queues - HIGH, MEDIUM, and LOW. Given a template, your message will reach within a certain SLO if sent through a certain queue
+1. A template can be of type PRIORITY, NORMAL, and BULK which ***currently*** maps directly to the HIGH, MEDIUM, and LOW queues. 
+1. The templates refer to both SMS and Email templates, but for shortcodes, this ADR focuses on SMS use cases
+
+# Implementation
+
+## Template Categorization
+
+When a client creates a template we want to be able to Categorize it into a certain type of template. We will start with the following Template Categories that were created through research on Notify's current templates.
+
+| Template Category | Example  |
+| ----------------- | -------- |
+| Status Update | Notice of change in status<br>The progress of a submission |
+| <br>Promotional call to action | Surveys<br>General apply now<br>"Learn more" |
+| Service-related requests | Submit additional documents<br>Follow up to move a process forward |
+| Fulfillment with attachments - email only | "Here’s your permit" |
+| <br>Broadcast messages | General information - (not related to transactions such as COVID-19 information)<br>Newsletters<br>Status updates |
+| <br>Auto-reply | No-reply messages<br>Acknowledgments<br>Response wait times |
+| Verification message | Authentication codes <br>Confirmation of account changes |
+| Confirmation / Receipts | Record of transactions<br>Approvals |
+| Other | User defined |
```

Comment:
Do we have system status content in SMS? I guess that is mostly around email notifications?

Reply:
Some GCNotify users send email notifications that are basically system status around their own internal system, i.e. _The system failed latest status check_. They integrate with our solution to have a quick email solution. I was wondering if we had such usage for SMS (as we do for emails) because that would be a category of its own.

Reply:
Yes, we do have SMS templates for internal system alerts and monitoring. That category would fall under "Alerts and Monitoring" or some similar wording, and then message type would be SMS.